test/testtokenrange.cpp

```diff
@@ -0,0 +1,108 @@
+/*
+ * Cppcheck - A tool for static C/C++ code analysis
+ * Copyright (C) 2007-2020 Cppcheck team.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include "settings.h"
+#include "testsuite.h"
+#include "testutils.h"
+#include "token.h"
+#include "tokenize.h"
+#include "tokenlist.h"
+#include "tokenrange.h"
+#include "symboldatabase.h"
+
+#include <string>
+#include <vector>
+
+struct InternalError;
+
+
+class TestTokenRange : public TestFixture {
+public:
+    TestTokenRange() : TestFixture("TestTokenRange") {
+    }
+
+private:
+    void run() OVERRIDE {
+        TEST_CASE(enumerationToEnd);
+        TEST_CASE(toEndHelper);
+        TEST_CASE(partialEnumeration);
+        TEST_CASE(scopeExample);
+        TEST_CASE(exampleAlgorithms);
+    }
+
+    void enumerationToEnd() const {
+        givenACodeSampleToTokenize var("void a(){} void main(){ if(true){a();} }");
+        const Token* expected = var.tokens();
+        for (auto t : ConstTokenRange{var.tokens(), nullptr }) {
+            ASSERT_EQUALS(expected, t);
+            expected = t->next();
+        }
+        const Token* const terminator = nullptr;
+        ASSERT_EQUALS(terminator, expected);
+    }
+
+    void toEndHelper() const {
+        givenACodeSampleToTokenize var("void a(){} void main(){ if(true){a();} }");
+        const Token* expected = var.tokens();
+        for (auto t : var.tokens()->toEnd()) {
+            ASSERT_EQUALS(expected, t);
+            expected = t->next();
+        }
+        const Token* const terminator = nullptr;
+        ASSERT_EQUALS(terminator, expected);
+    }
+
+    void partialEnumeration() const {
+        givenACodeSampleToTokenize var("void a(){} void main(){ if(true){a();} }");
+        const Token* expected = var.tokens()->tokAt(4);
+        const Token* end = var.tokens()->tokAt(10);
+        for (auto t : ConstTokenRange{ expected, end }) {
+            ASSERT_EQUALS(expected, t);
```

Comment:
I am skeptic about comparing pointers with ASSERT_EQUALS. Imagine that the assertion fails. Then I fear that the output will look like this:

   expected: 0x12345678 actual: 0x12345670

.. no way to see which token the expected and actual are .. just that they are different. The ASSERT might be more proper.

also I believe it is unfortunate to put an assertion in a for loop body. if it fails we will see what line it fails on but not how far the testing went before there was an assertion failure. it will not be informative.

what do you think about something like:

    const Token *start = var.tokens()->tokAt(4);
    const Token *end = var.tokens()->tokAt(10);
    int count = 0;
    for (const Token *tok: ConstTokenRange{start,end})
    {
        if (tok != start->tokAt(count))
            break;
        count ++;
    }
    ASSERT_EQUALS(6, count);

Reply:
I don't think it's any worse. I'm usually not fussed about assertion messages because I figure when a test fails you're going ot debug it anyway. I do feel strongly that we should be comparing the pointers (rather than say strings) but your solution does do that.

Reply:
there is some truth to that. you probably want to debug it also. but well the assertion output can be pretty useful. please change it..

Reply:
I had a play with this. I'm not super happy about having to hard code the number of tokens because it's less flexible. We're trying to test the Token class not the tokenizer and there are some situations where I think the tokenizer gets it wrong which I'd like to be able to fix without updating unrelated tests like this. What are your thoughts on this alternative which I think  achieves the best of both worlds?

```
    void enumerationToEnd() const {
        givenACodeSampleToTokenize var("void a(){} void main(){ if(true){a();} }");
        const Token* expected = var.tokens();
        int correctTokens = 0;
        for (auto t : ConstTokenRange{var.tokens(), nullptr }) {
            std::ostringstream message;
            message << "Token mismatch on token " << correctTokens;
            ASSERT_EQUALS_MSG(expected, t, message.str());
            expected = t->next();
        }
        const Token* const terminator = nullptr;
        ASSERT_EQUALS(terminator, expected);
    }
```

Reply:
It seems we should avoid `Tokenizer` in this test. Could you try out `TokenList` instead? Something like:

        std::istringstream istr(code);
        TokenList tokenList(&settings);
        tokenList.createTokens(istr, "test.cpp");

        const Token *startToken = tokenList.front()->tokAt(123);
        ....

Reply:
I guess we want to refactor the test also.. create some kind of test function. For instance it could return some error message. Maybe `testTokenRange(code,start,end)` is a good signature. And then we can write a test more or less like:

    ASSERT_EQUALS("", testTokenRange("your code", 1, 10))

Reply:
breaking it out is maybe not required.. I do not know if there are a lot of variations to test. maybe 1-3 tests are enough.

Reply:
I agree it makes sense to use TokenList rather than Tokenizer to generate the tokens because that minimizes the dependencies of the test. I still wouldn't want to manually count the tokens though.

I think it's probably worth  refactoring out the check as you suggest.

Reply:
actually I have to use the tokenizer for the test involving scopes because I need the symbol database to generate the scopes

Reply:
take a look at the latest incarnation. The first three tests are modified in the way you suggest (more or less). The last two are sort of different so the suggestions don't apply as far as I can see.